File: 0289-game-of-life/0289-game-of-life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        
        output = []
        
        m = len(board)
        n = len(board[0])
        
        def neighbours(i,j):
            direction = [(0,-1), (0,1),(1,0),(-1,0),(1,-1),(-1,1),(1,1),(-1,-1)]
            nei = 0
            for x,y  in direction:
                    x_cor = i + x
                    y_cor = j + y
                    if 0<= x_cor < m and 0<= y_cor < n:
                        if board[x_cor][y_cor] == 1 or board[x_cor][y_cor] == 3:
                            nei +=1
            return nei
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == 0:
                    if neighbours(i,j) == 3:
                        board[i][j] = 2
                    
                elif board[i][j] == 1:
                    if neighbours(i,j) not in [2,3]:
                        board[i][j] = 3
            
        
        for i in range(m):
            for j in range(n):
                if board[i][j]  == 2:
                    board[i][j] = 1
                elif board[i][j]  == 3:
                    board[i][j] = 0
        
        
    

File: 0289-game-of-life/0289-game-of-life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int m = board.length;
        int n = board[0].length;

        for(int i =0; i < m; i++){
            for (int j = 0; j < n; j++){
                int nei = neighbours(i,j,m,n,board);
                if (board[i][j] == 0){
                    if (nei == 3){
                        board[i][j] = 3; // live == 1
                    }
                }
                if (board[i][j] == 1){
                    if (nei != 2 && nei != 3){
                        board[i][j] = 2; // dead == 0
                    }
                }
            }
        }

        for(int i =0; i < m; i++){
            for (int j = 0; j < n; j++){
                if (board[i][j] == 3){
                    board[i][j] = 1;
                }
                if (board[i][j] == 2){
                    board[i][j] = 0;
                }
            }
        }
    }

    public int neighbours(int r, int c, int m, int n, int[][] board){
        int[][] dir = {{0,-1}, {0,1},{1,0},{-1,0},{1,-1},{-1,1},{1,1},{-1,-1}};
        int nei = 0;
        for(int i =0; i< dir.length; i++){
            int r1 = r + dir[i][0];
            int c1 = c + dir[i][1];

            if (r1>=0 && r1<m && c1>=0 && c1<n && (board[r1][c1] == 1 || board[r1][c1] == 2)){
                nei += 1;
            }
        }
        return nei;
    }
}